src/jvm.rs

```diff
@@ -72,45 +74,45 @@ pub trait JvmOp: Copy {
         AsUpcast::new(self)
     }
 
-    /// Given a JVM op that creates a local reference, convert the local reference
-    /// into a global one. Global JVM references can be held as long as you like
-    /// within
-    fn global(self) -> GlobalOp<Self>
-    where
-        for<'jvm> <Self as JvmOp>::Output<'jvm>: IntoGlobal<'jvm>,
-    {
-        GlobalOp::new(self)
-    }
-
     fn catch<J>(self) -> TryCatch<Self, J>
     where
         J: Upcast<Throwable>,
     {
         TryCatch::new(self)
     }
 
-    /// Given a JVM op that returns some Java type, convert it to its Rust equivalent
-    /// (e.g., from a Java String to a Rust string).
-    fn to_rust<R>(self) -> ToRustOp<Self, R>
+    /// Execute on the JVM, starting a JVM instance if necessary.
+    ///
+    /// Depending on the type parameter `R`,
+    /// this method can either return a handle to a Java object
+    /// or a Rust type:
+    ///
+    /// * When `R` is something like [`Java<java::lang::String>`][`Java`],
+    ///   this method will return a handle to a Java object.
+    ///   Note that to account for possible null return values
+    ///   you may need to either invoke `assert_not_null` or else
+    ///   use a result type with an `Option`, e.g.,
+    ///   `Option<Java<java::lang::String>>`.
+    /// * When `R` is a Rust type like [`String`][],
+    ///   this method will convert the Java value into a Rust type.
+    ///   You may need to derive `ToRust` for your Rust type
+    ///   to indicate how the Java object is to be converted.
+    fn execute<R>(self) -> crate::Result<R>
     where
         for<'jvm> Self::Output<'jvm>: IntoRust<R>,
     {
-        ToRustOp::new(self)
+        Jvm::with(|jvm| self.execute_with(jvm))
     }
 
-    /// Execute the jvm op, starting a JVM instance if necessary.
-    /// To use this method, the result type cannot be tied to the JVM.
-    /// Typically this is achieved by a call to [`to_rust()`][`Self::to_rust`],
-    /// but if you wish to hold on to a reference to a JVM object,
-    /// you can use [`global()`][`Self::global`] to create a global reference.
-    fn execute<R>(self) -> crate::GlobalResult<R>
+    /// Internal method
+    fn execute_with<'jvm, R>(self, jvm: &mut Jvm<'jvm>) -> crate::LocalResult<'jvm, R>
     where
-        for<'jvm> Self: JvmOp<Output<'jvm> = R>,
+        for<'j> Self::Output<'j>: IntoRust<R>,
     {
-        Jvm::with(|jvm| self.execute_with(jvm))
+        ToRustOp::new(self).do_jni(jvm)
     }
-
-    fn execute_with<'jvm>(self, jvm: &mut Jvm<'jvm>) -> crate::Result<'jvm, Self::Output<'jvm>>;
+    /// Internal method
```

Comment:
a slightly more detailed comment would probably be useful for folks working on the project since this is now the fundamental "do a thing" message

Reply:
since `execute_with` converts into `Rust` vs. `do_jni` which keeps it in Java maybe we could rename `execute_with` something that mentioned the conversion into Rust